### DISCUSSION TITLE
Code library for IPSEC AH

### DISCUSSION THREAD

**Discussion by huaraz** [strongswan/strongswan#2152] [Q&A] [answered]
Hi Team,

   I am new to strongswan and wonder where i.e. which library / file I can find the code section for AH packet creation / verification ?

Thank you
Markus

**Comment by tobiasbrunner** [comment, marked as answer]
You might want to start with the [introduction](https://docs.strongswan.org/docs/5.9/howtos/introduction.html). You'd learn that IPsec traffic is not usually handled by strongSwan, but the operating system (i.e. for the implementation check the Linux kernel source code). While we do have a userland ESP implementation (for mobile clients), there is nothing similar for AH (which is rarely used anyway and could be tricky to implement in userland).

**Comment by huaraz** [comment]
Thank you for the clarification. I interpreted the below that complete meant it includes kernel modules.



_**Securing a Network**
strongSwan is a complete IPSEC solution providing encryption and authentication to servers and clients. strongSwan can be used to secure communications with remote networks, so that connecting remotely is the same as connecting locally._
